src/Data/BitVector/Sized/Internal.hs

```diff
@@ -1,419 +1,748 @@
-
--- | Right logical shift.
-bvShiftRL :: BitVector w -> Int -> BitVector w
-bvShiftRL bv@(BV wRepr _) shf = BV wRepr (truncBits width (x `shift` (- toPos shf)))
-  where width = natValue wRepr
-        x     = bvIntegerU bv
-
--- | Bitwise rotate.
-bvRotate :: BitVector w -> Int -> BitVector w
-bvRotate bv rot' = leftChunk `bvOr` rightChunk
-  where rot = rot' `mod` bvWidth bv
-        leftChunk = bvShift bv rot
-        rightChunk = bvShift bv (rot - bvWidth bv)
-
--- | Get the width of a 'BitVector'.
-bvWidth :: BitVector w -> Int
-bvWidth (BV wRepr _) = fromIntegral (natValue wRepr)
+complement :: NatRepr w -> BV w -> BV w
+complement w (BV x) = mkBV' w (B.complement x)
+
+-- | Left shift by positive 'Natural'.
+shl :: NatRepr w -> BV w -> Natural -> BV w
+shl w (BV x) shf = checkNatural shf $
+  mkBV' w (x `B.shiftL` fromIntegral shf)
+
+-- | Right arithmetic shift by positive 'Natural'.
+ashr :: NatRepr w -> BV w -> Natural -> BV w
+ashr w bv shf = checkNatural shf $
+  mkBV' w (asSigned w bv `B.shiftR` fromIntegral shf)
+
+-- | Right logical shift by positive 'Natural'.
+lshr :: BV w -> Natural -> BV w
+lshr (BV x) shf = checkNatural shf $
+  -- Shift right (logical by default since the value is positive). No
+  -- need to truncate bits, since the result is guaranteed to occupy
+  -- no more bits than the input.
+  BV (x `B.shiftR` fromIntegral shf)
+
+-- | Bitwise rotate left.
+rotateL :: NatRepr w -> BV w -> Natural -> BV w
+rotateL w bv rot' = leftChunk `or` rightChunk
+  where rot = rot' `mod` wNatural
+        leftChunk = shl w bv rot
+        rightChunk = lshr bv (wNatural - rot)
+        wNatural = natValue w
+
+-- | Bitwise rotate right.
+rotateR :: NatRepr w -> BV w -> Natural -> BV w
+rotateR w bv rot' = leftChunk `or` rightChunk
+  where rot = rot' `mod` wNatural
+        rightChunk = lshr bv rot
+        leftChunk = shl w bv (wNatural - rot)
+        wNatural = natValue w
+
+-- | The zero bitvector of any width.
+zero :: NatRepr w -> BV w
+zero w = checkNatRepr w $ BV 0
+
+-- | The bitvector with value 1, of any positive width.
+one :: 1 <= w => NatRepr w -> BV w
+one w = checkNatRepr w $ BV 1
+
+-- | The bitvector whose value is its own width, of any width.
+width :: NatRepr w -> BV w
+width w = checkNatRepr w $ BV (intValue w)
+
+-- | The 'BV' that has a particular bit set, and is 0 everywhere
+-- else.
+bit :: ix+1 <= w
+    => NatRepr w
+    -- ^ Desired output width
+    -> NatRepr ix
+    -- ^ Index of bit to set
+    -> BV w
+bit w ix = checkNatRepr w $ BV (B.bit (widthVal ix))
+
+-- | Like 'bit', but without the requirement that the index bit refers
+-- to an actual bit in the output 'BV'. If it is out of range, just
+-- silently return the zero bitvector.
+bit' :: NatRepr w
+     -- ^ Desired output width
+     -> Natural
+     -- ^ Index of bit to set
+     -> BV w
+bit' w ix = mkBV w (B.bit (fromIntegral ix))
+
+-- | @setBit w bv ix@ is the same as @or bv (bit w ix)@.
+setBit :: ix+1 <= w
+       => NatRepr w
```

Comment:
Is this `NatRepr` just for symmetry?  I don't want to bikeshed on it, just noting that `w` is already fixed by the original `BV`. I think the symmetry argument is fine if that is your intent.

Reply:
I removed the width arg -- I really only want to ask for a `NatRepr` arg when we either need it at runtime in order to compute the function, or when we need it at compile time to determine the return type.